bin/test_hypothesis.py

```diff
@@ -0,0 +1,256 @@
+            ]
+        ),
+        x=expr,
+        y=expr,
+    )
+    def add_expr_ops(self, op, x, y):
+        return op(x, y)
+
+    @rule(
+        target=expr,
+        fn=st.sampled_from(
+            [
+                sympy.sin,
+                sympy.cos,
+                sympy.tan,
+                sympy.asin,
+                sympy.acos,
+                sympy.atan,
+                sympy.sqrt,
+                sympy.powsimp,
+                sympy.expand_power_exp,
+                sympy.expand_power_base,
+                sympy.powdenest,
+                sympy.expand_log,
+                sympy.logcombine,
+                sympy.expand_func,
+                sympy.combsimp,
+                sympy.trigsimp,
+                sympy.expand_trig,
+                sympy.factorial,
+            ]
+        ),
+        x=expr,
+    )
+    def fn_expr(self, fn, x):
+        return fn(x)
+
+    @rule(target=expr, x=expr, y=expr, z=expr)
+    def subs_expr(self, x, y, z):
+        return x.subs(y, z)
+
+    @rule(target=expr, x=expr)
+    def simplify_expr(self, x):
+        return sympy.simplify(x)
+
+    @rule(target=expr, x=expr)
+    def cancel_expr(self, x):
+        return sympy.cancel(x)
+
+    @rule(target=expr, x=expr, force=st.booleans())
+    def expand_expr(self, x, force):
+        return sympy.expand(x, force=force)
+
+    @rule(target=expr, x=expr)
+    def factor_expr(self, x):
+        try:
+            return sympy.factor(x)
+        except ValueError:
+            reject()
+
+    @rule(target=expr, x=expr)
+    def apart_expr(self, x):
+        try:
+            return sympy.apart(x)
+        except (sympy.PolynomialError, sympy.NotAlgebraic, NotImplementedError):
+            reject()
+
+    @rule(target=expr, x=expr, y=expr)
+    def collect(self, x, y):
+        return sympy.collect(x, y)
+
+    @rule(x=expr)
+    def expr_doit(self, x):
+        x.doit()
+
+    @rule(x=expr)
+    def expr_evalf(self, x):
+        x.evalf()
+
+    @rule(
+        target=expr,
+        x=expr,
+        var=v,
+        f=c,
+        t=c,
+        op=st.sampled_from([sympy.Sum, sympy.Product]),
+    )
+    def combine_many(self, x, var, f, t, op):
+        return op(x, (var, f, t))
+
+    @rule(source=expr)
+    def doit_is_idempotent(self, source):
+        done = source.doit()
+        if done == done:
+            assert done == done.doit()
+
+    @rule(source=expr, data=st.data())
+    def substitution_commutes_with_simplify(self, source, data):
+        symbols = sorted(
+            [v for v in source.atoms() if isinstance(v, sympy.Symbol)],
+            key=lambda a: a.name,
+        )
+
+        if not symbols:
+            reject()
+
+        variables = data.draw(st.permutations(symbols), label="variables")
+        assignment = [sympy.Integer(data.draw(st.integers())) for _ in variables]
+
+        r1 = source.subs(dict(zip(variables, assignment)))
+
+        if isinstance(r1, sympy.numbers.NaN) or r1 != r1:
```

Comment:
All SymPy objects including sympy.nan compares equal with themselves. If they don't that would be a bug.